velox/connectors/hive/PartitionIdGenerator.cpp

```diff
@@ -0,0 +1,139 @@
+    hashers_.push_back(exec::VectorHasher::create(
+        inputType->childAt(partitionChannels_[i]), partitionChannels_[i]));
+  }
+}
+
+std::map<uint64_t, uint64_t> PartitionIdGenerator::run(
+    const RowVectorPtr& input,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+  bool rehashed =
+      computeIds(input, SelectivityVector{input->size(), true}, result);
+
+  std::map<uint64_t, uint64_t> idMap;
+  if (rehashed && allPartitionsVector_->size() > 0) {
+    raw_vector<uint64_t> newIds(allPartitionsVector_->size());
+
+    VELOX_CHECK(
+        !computeIds(allPartitionsVector_, activeIds_, newIds),
+        "Partition ID generator should not rehash when computing for the existing partitions.");
+
+    activeIds_.applyToSelected([&](auto id) { idMap[id] = newIds[id]; });
+
+    movePartitions(idMap);
+  }
+
+  addPartitions(input, result);
+
+  return idMap;
+}
+
+bool PartitionIdGenerator::computeIds(
+    const RowVectorPtr& input,
+    const SelectivityVector& rows,
+    raw_vector<uint64_t>& result) {
+  result.resize(input->size());
+
+  for (auto& hasher : hashers_) {
+    auto partitionVector = input->childAt(hasher->channel())->loadedVector();
+    hasher->decode(*partitionVector, rows);
+  }
+
+  bool rehashed = false, toRehash = false;
+  do {
+    toRehash = false;
+    for (auto& hasher : hashers_) {
+      if (!hasher->computeValueIds(rows, result)) {
+        toRehash = true;
+      }
+    }
+    if (toRehash) {
+      rehashed = true;
+
+      uint64_t multiplier = 1;
+      for (auto i = 0; i < hashers_.size(); i++) {
+        multiplier = hashers_[i]->enableValueIds(multiplier, kHasherReservePct);
+        VELOX_CHECK_NE(
+            multiplier,
+            exec::VectorHasher::kRangeTooLarge,
+            "Exceeded limit of distinct partitions.");
+      }
+    }
+  } while (toRehash);
+
+  return rehashed;
+}
+
+void PartitionIdGenerator::movePartitions(
+    const std::map<uint64_t, uint64_t>& indexMap) {
+  vector_size_t maxNewId = indexMap.rbegin()->second;
+
+  allPartitionsVector_->resize(maxNewId + 1, false);
+  activeIds_.resize(maxNewId + 1, false);
+
+  for (auto entry = indexMap.rbegin(); entry != indexMap.rend(); entry++) {
+    allPartitionsVector_->copy(
+        allPartitionsVector_.get(), entry->second, entry->first, 1);
+    allPartitionsVector_->setNull(entry->first, true);
+    activeIds_.setValid(entry->second, true);
+    activeIds_.setValid(entry->first, false);
+  }
+  activeIds_.updateBounds();
+}
+
+void PartitionIdGenerator::addPartitions(
+    const RowVectorPtr& input,
+    const raw_vector<uint64_t>& inputIds) {
+  maxId_ = *std::max_element(inputIds.begin(), inputIds.end());
+  if (maxId_ > allPartitionsVector_->size() - 1) {
+    allPartitionsVector_->ensureWritable(SelectivityVector::empty(maxId_ + 1));
+    activeIds_.resize(maxId_ + 1, false);
+  }
+
+  for (vector_size_t row = 0; row < inputIds.size(); row++) {
```

Comment:
Do we need to run this loop if maxId hasn't increased?

Reply:
I did once think about it, and, yes, even current input doesn't increase history max, there still could be new combinations of partition col values and new IDs correspondingly that are smaller than history max. And need to copy and track the partition values here.

Ex., suppose we have two partition keys p1 and p2. And the two VectorHashers already contain p1=0/p2=0 and p1=2/p2=2. A new input could have p1=0/p2=2.